job-shop-scheduling/demo.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2019 D-Wave Systems Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+from dwave.system.composites import EmbeddingComposite
+from dwave.system.samplers import DWaveSampler
+
+from job_shop_scheduler import get_jss_bqm
+
+# Construct a BQM for jobs 'a', 'b' and 'c'
+jobs = {"a": [("mixer", 2), ("oven", 1)],
+        "b": [("mixer", 1)],
+        "c": [("oven", 2)]}
+max_time = 4	  # Upperbound on how long the schedule can be
```

Comment:
This seems to come out of nowhere here... it is not based on the four defined constraints above. Can you say something about why you set this restriction and where its value comes from?

Reply:
I have added the following to `README.rst` to explain `max_time`.

```
In ``demo.py`` and ``job_shop_scheduler.py``, we see a variable called
``max_time``. It refers to the maximum possible end time in our job shop
schedule.

Naively, we could set our ``max_time`` to infinity, so that our solver
would consider all possible schedules with end times from 0 to infinity.
However, this is a huge space to explore, and makes our BQM unnecessarily
large and difficult to solve.

Instead, we can apply our knowledge on the *worst possible* schedule scenario
so that we can put an upper bound on the schedule end times. The worst possible scenario
is if all job tasks require the same exact machine, hence there is no opportunity
for parallelization. In this case, the schedule end time is the sum of all
task durations because that one machine will run those tasks back-to-back. We know
the optimal schedule for these tasks must finish earlier or at the same time as
this worst case scenario because these tasks don't necessarily all need to run
on that one machine; this allows for parallelization and a shorter schedule. Thus,
by default, the ``max_time`` considered for a schedule is the sum of task durations.

Note that we can lower ``max_time`` so that the solver considers a smaller space
of schedule solutions. In terms of quantum computing hardware, this means
using fewer qubits as we are considering over a smaller range of end times
and thus, fewer possible schedules. This is acceptable so long as the optimal
schedule has an end time that is less than ``max_time``. Otherwise, no valid
schedule would be explored as we are considering schedule end times that are
shorter than that of the optimal schedule.
```